src/board/helpers.mjs

```diff
@@ -0,0 +1,114 @@
+/* eslint-disable guard-for-in */
+/* eslint-disable no-restricted-syntax */
+
+// contains functions to duplicate, flatten, unflatten, and shuffle the grid
+
+function duplicateGrid(grid) {
+  const result = [];
+  grid.forEach((row, gridIndex) => {
+    const newRow = [];
+    row.forEach((color, rowIndex) => {
+      newRow[rowIndex] = color;
+    });
+    result[gridIndex] = newRow;
+  });
+  return result;
+}
+
+function flattenGrid(grid) {
+  const result = [];
+  grid.forEach((row) => {
+    row.forEach((color) => {
+      result.push(color);
+    });
+  });
+  return result;
+}
+
+function shuffle(arr) { // Fisher-Yates shuffle
+  for (let i = arr.length - 1; i > 0; i -= 1) {
+    const j = Math.floor(Math.random() * (i + 1));
+    // TODO: ask about this
+    // eslint-disable-next-line no-param-reassign
+    [arr[i], arr[j]] = [arr[j], arr[i]];
+  }
+  return arr;
+}
+
+function flatToMultiDimensional(flatGrid) {
+  // square root of the size of the grid will tell us how many rows/columns
+  const rowSize = Math.sqrt(flatGrid.length);
+
+  // create the empty multidimensional grid
+  const multiDimensionalGrid = new Array(rowSize);
+
+  // why can I not turn this into a forEach or map
+  for (let i = 0; i < multiDimensionalGrid.length; i += 1) {
+    multiDimensionalGrid[i] = new Array(rowSize);
+  }
+
+  // use these variables to keep track of where in our md array we are as we iterate
+  let outerIndex = 0;
+  let innerIndex = 0;
+  // now iterate through the flatGrid and put all its items into the md grid
+  for (let i = 0; i < flatGrid.length; i += 1) {
+    multiDimensionalGrid[outerIndex][innerIndex] = flatGrid[i];
+    innerIndex += 1;
+    if (innerIndex % rowSize === 0) {
+      outerIndex += 1;
+      innerIndex = 0;
+    }
+  }
+
+  // but why is this angry about a lack of return statement
+  // flatGrid.map((color) => {
```

Comment:
maps return a new array with each element thats a result of the callback function, so if there is a conditional, that wont let it return correctly all of the time, so you might need to add `return` outside of the if statement, like right after it. Thats my best guess just looking at the code

Reply:
That's good to know. I think I'm going to keep it as a for loop right now--do you think it's possible to always find a way to replace a for loop with a map/forEach? I have quite a few for loops in other parts of my code that I wasn't able to figure out, which might be a nice thing for me to go back and look at sometime.